src/dashboard/components/list_component.py

```diff
@@ -0,0 +1,113 @@
+"""List component."""
+
+from typing import List, Tuple
+
+from dash import dcc, html
+
+from dashboard.components.icon import icon
+
+
+def generate_row_item(content: str) -> html.Span:
+    """Generate a list row item.
+
+    Args:
+        content (str): The content to display inside the row item.
+
+    Returns:
+        Component: The list row item.
+    """
+    return html.Span(
+        className=(
+            "last:border-r-0 flex-1 border-r-2 pl-2 py-1 text-ellipsis overflow-hidden"
+            " whitespace-nowrap"
+        ),
+        children=content,
+    )
+
+
+def generate_list_row(list_row_data: Tuple[int, List[str]]) -> dcc.Link:
+    """Generate a list row.
+
+    Args:
+        list_row_data (Tuple[int, List[str]]): The row index and
+        content of the row.
+
+    Returns:
+        Component: The list row.
+    """
+    index = list_row_data[0]
+    list_row = list_row_data[1]
+    return dcc.Link(
+        id={"type": "list-row", "index": index},
+        href="/",
+        className=(
+            "flex pl-2 justify-start items-center border-b-2 border-gray-400"
+            " text-base cursor-pointer hover:bg-gray-100"
+        ),
+        children=[icon("analytics", fill=1, size=28)] + list(map(generate_row_item, list_row)),
+    )
+
+
+def generate_list_titles(list_titles: List[str]) -> List[html.Span]:
+    """Generate list title elements.
+
+    Args:
+        list_titles (List[str]): A list of titles.
+
+    Returns:
+        List[html.Div]: A list of title elements.
+    """
+    return list(map(generate_row_item, list_titles))
+
+
+def generate_list_rows(list_rows: List[List[str]]) -> List[dcc.Link]:
+    """Generate list row elements.
+
+    Args:
+        list_rows (List[List[str]]): A list of rows.
+
+    Returns:
+        list[html.Div]: A list of row elements.
+    """
+    return list(map(generate_list_row, enumerate(list_rows)))
+
+
+def list_component(list_titles: List[str], list_rows: List[List[str]], _id: str) -> html.Div:
+    """Create a list component.
+
+    Args:
+        titles_names (List[str]): The titles to display at the top of
+        the list.
+        list_rows (List[List[str]]): The rows that make up the list
+        contents.
+
+    Raises:
+        IndexError: If the amount of titles does not match the amount
+        of list rows.
+
+    Returns:
+        html.Div: A list component.
+    """
+    if len(list_rows) > 0 and not all(len(list_titles) == len(list_row) for list_row in list_rows):
```

Comment:
This is difficult to read. I feel like a more suitable data structure could solve some of the headaches in this function. How about `list[tuple[str, list[str]]]`?

Reply:
This is likely a more suitable data structure, but what if the length of one `list[str]` column in the `list[tuple[str, list[str]]]` object has a different length than the others? We'd be stuck with the same headache.